### ISSUE TITLE
SV Crashes when Clicking on Segmentations Generated from Python Script

### ISSUE THREAD

**Issue by s3nming** [SimVascular/SimVascular#1029]
This is a crashing bug(and error message display) . After loading in a few python scripts into SV, an error message would display, saying:
"An error occurred. You should save all data and quit the program to prevent possible data loss"
Usually, if you try to click on anything else after hitting the 'Ignore' button, this message would be spammed on your screen before SimVascular crashes.

**To Reproduce**
1. Create a new project
2. Click on 'Python Console'
3. Click on 'Text Editor'
4. Click on the Folder Icon(first icon at the top)
5. Load in Python Script(below)
Script:
[PA.py.zip](https://github.com/SimVascular/SimVascular/files/12176213/PA.py.zip)
7. Click on the run icon
8. Repeat from step 2 but with another Python Script(Below)
Script:
[corrected_exponential_down_right.py.zip](https://github.com/SimVascular/SimVascular/files/12176242/corrected_exponential_down_right.py.zip)
9. Double click on Segmentation(either one of the subsections) on the left hand side
10. If the error does not show up, click the middle box in the top right of the 'Display Screen'.
11. Click 'Ignore'
12. Click on anything(E.G: Pathing)

**Expected behavior**
I expect the idealized model to appear on the Display screen, but instead error messages and crashes occur.

**Screenshots**
First error message:

<img width="1438" alt="Screen Shot 2023-07-26 at 11 19 34 AM" src="https://github.com/SimVascular/SimVascular/assets/130249623/770768f9-ca8b-4e80-b8d4-56713eddca44">
Bugged-out SV Screen:

![Screen Shot 2023-07-26 at 11 40 46 AM](https://github.com/SimVascular/SimVascular/assets/130249623/cabfff28-7e3b-4606-b722-11d705b0f0c5)

Spammed error message:

<img width="1439" alt="Screen Shot 2023-07-26 at 11 51 55 AM" src="https://github.com/SimVascular/SimVascular/assets/130249623/3be42b50-18b8-4b1d-b2e9-fc6d26ceac80">

**Desktop:**
 - OS: macOS Monterey
 - SimVascular Version 2023-05-23

**Additional context**
This bug randomly showed up, despite the fact that I had not changed anything about my scripts or SV project. About 4 weeks ago, I opened up SV, created a new project, and started to loading in my python code to the software. Then this bug showed up out of nowhere. I also encountered this bug about 7-6 weeks ago, but it disappeared on its own. It showed up again 4 weeks ago.

**Comment by JonathanPham** [comment]
I updated the `corrected_exponential_down_right.py.zip` script so that the `helper_functions` Python module is removed. Attached is a standalone, working script. 

I was unable to reproduce this issue on SimVascular (version 2021.10.03, compiled from the GitHub source) on my Ubuntu desktop.

[corrected_exponential_down_right.zip](https://github.com/SimVascular/SimVascular/files/12176418/corrected_exponential_down_right.zip)

**Comment by zasexton** [comment]
Yes, this is an error that I have run into before... reducing the number of contours per vessel has allowed me to more reliably exchange segmentation data between the python interface and the data manager.

**Comment by ktbolt** [comment]
I followed the above instructions and reproduced the failure. Selecting the `Show Details` button on the error popup shows
```
/usr/local/sv/ext/2019.06/release/gl2/src/mitk-2018.04.2/Modules/Core/src/DataManagement/mitkBaseGeometry.cpp:460:
itk::ERROR: PlaneGeometry(0x7fddb6653ee0): Internal ITK matrix inversion error, cannot proceed. Matrix was: 
0 nan 0
0 nan 0
0 nan 0
Suggested inverted matrix is:
nan nan nan
nan nan nan
nan nan nan
```

The problem is that there is no image data for the project so there is no valid transformation to image coordinates.

**Comment by ktbolt** [comment]
SV is an  image-based modeling  app so what you are trying to do is outside of what it was implemented for.

@s3nming Why are you even trying to do this within the SV GUI with no image context? This should be done just using the SV Python API without the GUI.

**Comment by zasexton** [comment]
@ktbolt I'm confused why the Python GUI would ever be successful if no valid transformation would ever exist? It is not always the case that SV crashes despite never having image data when working with the Python API in the GUI

**Comment by ktbolt** [comment]
@zasexton I think it depends on what you are doing in the GUI. The image slicing uses the transformation information.

Or maybe the real problem is that the Python API `dmg` module is freeing data when it should not and therefore corrupting the executable data arena.

**Comment by JonathanPham** [comment]
@ktbolt We were using the SV GUI to quickly visualize the pathlines and segmentations that we generated. We also wanted to use the GUI so that if the segmentations generated from the Python script intersect each other, then we manually fix those problematic segmentations to get the lofting to work. 

Like @zasexton mentioned, sometimes when we click on a contour in the `Segmentations` tab, the ITK matrix inversion error does not appear and other times, it does.

**Comment by ktbolt** [comment]
The Python API `dmg` module looks fine. 

@JonathanPham You can visualize the pathlines and segmentations without the GUI. The scripts here https://github.com/SimVascular/SimVascular-Tests/tree/master/new-api-tests contains a simple graphics module you can use to display geometry.

The SV failure I believe is caused by the `SV 2D Segmentation Tool' 'Reslice' trying to extract a 2D slice from an image that is not there. The 2D slice is automatically extracted when clicking on a Segmentations data node. This seems to be causing undefined behavior in the C++ code. 

It would be better to work outside of SV to fix problematic segmentations, maybe create an interactive app to do this.

But if you want to use SV then a workaround would be to create an artificial image (VTI) larger than the segmentations you are creating.

**Comment by JonathanPham** [comment]
That make sense. Using an artificial VTI image would be the easiest approach to manually edit the segmentations in the SV GUI. If we were to edit them outside of the GUI, that is definitely doable too, but would require us to make an interactive app to do this, like you mentioned.

Do you have a script that can generate an artificial VTI file, say one with just black-colored pixels?

Thanks for looking into this, Dave.

**Comment by ktbolt** [comment]
I will have a look at the SV code to see about adding an warning for no image data.

**Comment by alisonmarsden** [comment]
﻿It seems SV should issue a warning in this situation to say “you are missing image data” or similar so we avoid a crash with no explanation.    I do think some users may want to generate idealized models without an associated image data set and then set up a simulation using the GUI.    Is there a way to give an option to generate a synthetic / blank image data set in this situation?


From: Dave Parker ***@***.***>
Date: Wednesday, July 26, 2023 at 2:01 PM
To: SimVascular/SimVascular ***@***.***>
Cc: Subscribed ***@***.***>
Subject: Re: [SimVascular/SimVascular] SV Crashes when Clicking on Segmentations Generated from Python Script (Issue #1029)

@zasexton<https://github.com/zasexton> I think it depends on what you are doing in the GUI. The image slicing uses the transformation information.

Or maybe the real problem is that the Python API dmg module is freeing data when it should not and therefore corrupting the executable data arena.

**Comment by ktbolt** [comment]
I've modified the `org.sv.gui.qt.segmentation` plugin code to not crash SV if there is no image data. I think the reslice slider is used to set the 2D images in the SV graphics view and with no image data that was causing a failure; not 100% sure about this since the code is totally undocumented.

Now when the Segmentation Tool is activated a warning popup is displayed and most of its controls are disabled.

<img width="1157" alt="Screen Shot 2023-07-27 at 3 26 03 PM" src="https://github.com/SimVascular/SimVascular/assets/23109458/00394595-2ea8-40d0-8da2-c0c599cfe5a2">


I've disabled all GUI controls except for lofting and deleting a contour. 
<img width="1266" alt="Screen Shot 2023-07-27 at 3 27 31 PM" src="https://github.com/SimVascular/SimVascular/assets/23109458/f14f7ec9-86cb-426d-9041-5b1cc5b11659">


I had wanted to be able to manually edit contours (e.g. set the radius of a circle contour) but no dice, even the editing of contour parameters is tied to reslicing image data it seems and will therefore crash SV.
